torch/csrc/generic/Tensor.cpp

```diff
@@ -907,23 +906,45 @@ static THIndexTensor* THPTensor_(_calculateLinearIndices)(
   }
   THLongStorage_free(indexerSize);
 
+#ifdef THC_GENERIC_FILE
+  // Call GPU kernel for index calculation
+  THCudaLongTensor *cudaIndices =
+    THCudaLongTensor_newWithSize1d(LIBRARY_STATE indexingElements);
+  long baseOffset = THTensor_(storageOffset)(LIBRARY_STATE indexed);
+
+  // Need to pass broadcast Tensors to API, pass NULL ptr for all empty
+  // (i.e. not-advanced indexed) dims
+  THCudaLongTensor *indexers[THTensor_(nDimension)(LIBRARY_STATE indexed.get())];
+
+  for (int i = 0; i < THTensor_(nDimension)(LIBRARY_STATE indexed.get()); ++i) {
+    bool adv = flattenedBroadcasters.find(i) != flattenedBroadcasters.end();
+    if (adv) {
+      THCudaLongTensor *bcIndices = THCudaLongTensor_newWithSize1d(
+        LIBRARY_STATE THLongTensor_nElement(flattenedBroadcasters[i].get()));
+      THCudaLongTensor_copyAsyncCPU(LIBRARY_STATE bcIndices, flattenedBroadcasters[i].get());
+      indexers[i] = bcIndices;
+    } else {
+      indexers[i] = NULL;
+    }
+  }
+
+  THTensor_(calculateAdvancedIndexingOffsets)(LIBRARY_STATE cudaIndices, indexed, baseOffset, indexers);
+
+  // Free the indexers
+  for (int i = 0; i < THTensor_(nDimension)(LIBRARY_STATE indexed.get()); ++i) {
+    if (indexers[i] != NULL) {
+      THCudaLongTensor_free(LIBRARY_STATE indexers[i]);
```

Comment:
I guess it would be better to allocate a single chunk of memory and use views onto it for each dim